[parser] Flag single unparenthesized generator expr with trailing comma in arguments.

abhijeetbodas2001 opened the pull request:
Fixes #17867

## Summary

The CPython parser does not allow generator expressions which are the sole arguments in an argument list to have a trailing comma.
With this change, we start flagging such instances.

## Test Plan

Added new inline tests.

abhijeetbodas2001 commented on crates/ruff_python_parser/src/parser/mod.rs:
```diff
@@ -659,6 +661,8 @@ impl<'src> Parser<'src> {
 
         self.recovery_context = saved_context;
+
+        trailing_comma_range
```
Instead of special-casing inside `parse_comma_separated_list` itself, I felt returning the `trailing_comma_range` to be simpler. But please let me know if this feels hacky.

ntBre replied:
This looks reasonable to me (and the change is quite small overall once I hid the whitespace diff), but I have two ideas here:
- Can we just return a `bool` instead of an `Option<TextRange>`? It looks like we only ever use the `Option` to call `is_some`, so we could just call `is_some` in the return here.
- Could we update the `recovery_context_kind` we pass in here so that we reuse the existing `OtherError` on line 657?

The second of these would prevent the need for the first and seems like a more robust approach, but I'm not really familiar with the available `RecoveryContextKind`s. If we can't get that to work, I think the current approach (with the `bool` modification) would be fine. @dhruvmanila would know better than me though!

abhijeetbodas2001 replied:
The context is `RecoveryContextKind::Arguments`, which alows trailing commas.

By updating `recovery_context_kind`, do you mean making it so that `OtherError` is raised even for `RecoveryContextKind::Arguments`? And later on replace that `OtherError` with the more specialized `DuplicateKeywordArgumentError` (or remove it entirely, if generators are not involved)?

abhijeetbodas2001 replied:
I addressed the first point and pushed here.
Happy to revert/change if we decide on the other way.

ntBre replied:
Sorry for the confusion, I pulled down the PR locally, and I think I understand what's going on better now. My suggestion was basically a worse wording of your idea to add special-casing inside of `parse_comma_separated_list`, but I see now that the proper error (`UnparenthesizedGeneratorExpression`) is emitted from `validate_arguments`, so I think your implementation totally makes sense.

abhijeetbodas2001 commented:
@ntBre could you please review?

ntBre commented on crates/ruff_python_parser/resources/inline/ok/args_unparenthesized_generator.py:
```diff
@@ -1 +1,2 @@
 sum(x for x in range(10))
+sum((x for x in range(10),)
```
I think this is the cause of the CI failure:
```suggestion
sum((x for x in range(10)),)
```

abhijeetbodas2001 replied:
Oops, looks like a `git add` miss on my end. Fixed.

ntBre commented on crates/ruff_python_parser/src/parser/expression.rs:
```diff
@@ -661,117 +661,120 @@ impl<'src> Parser<'src> {
-                        }
-                    };
-
+        let trailing_comma_range =
```
This is just a note to other potential reviewers, but I highly recommend hiding whitespace from the diff here. The only change in this function is capturing the value `trailing_comma_range`, which is now returned by `parse_comma_separated_list`.

abhijeetbodas2001 replied:
(I also did not know about this setting. Thank you!)

abhijeetbodas2001 commented:
Thanks for reviewing, Brent. I addressed all but one comments.
Please see the question at https://github.com/astral-sh/ruff/pull/17893#discussion_r2076077042

ntBre commented on crates/ruff_python_parser/src/parser/expression.rs:
```diff
@@ -2541,7 +2544,7 @@ impl<'src> Parser<'src> {
         }
 
-        if arguments.len() > 1 {
+        if (has_trailing_comma && arguments.len() == 1) || arguments.len() > 1 {
```
I believe we can simplify this to
```suggestion
        if has_trailing_comma || arguments.len() > 1 {
```

since we'll either have 0 arguments, in which case the loop is never entered and the conditional doesn't matter; 1 argument in which case only `has_trailing_comma` matters; or many arguments, in which case `arguments.len() > 1` applies.

ntBre commented on crates/ruff_python_parser/resources/inline/ok/args_unparenthesized_generator.py:
```diff
@@ -1 +1,2 @@
 sum(x for x in range(10))
+sum((x for x in range(10)),)
```
I guess one more nit while I'm here, this is slightly confusing with the name of the test (`unparenthesized_generator`), but it's also annoying to have to rename snapshots and such, so I'm okay with leaving it as is.

abhijeetbodas2001 replied:
Hmm, if premise is un-parenthesized expressions are bad, then I think it would be natural for the "ok" parts of the test to be good (ie, have parens)?

abhijeetbodas2001 replied:
I actually added one more inline test here. Not related to the changes in this PR specifically, but I felt the case of multiple arguments was missing from the "ok" section.

ntBre reviewed (approved):
Thanks, this looks good to me! And thanks for clearing up my confusion from earlier. I found a couple of minor nits, but otherwise I just want to give Dhruv a chance to look at this if he wants to.

abhijeetbodas2001 commented:
Thanks Brent. I've addressed the review comments. For the test name -- IMO the name isn't exactly incorrect, so I'd prefer to keep it the same way.
(Although I did add a new inline test in the "ok" section which you might want to review.)

dhruvmanila reviewed (approved):
Thank you!
